### PR TITLE
Telemetry aggregated dispatcher

### PR DESCRIPTION
**Pull request by antrix1989**

### CODE REVIEW

File: MSAL/src/public/MSALTelemetryEventsObserving.h [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+// THE SOFTWARE.
+//
+//------------------------------------------------------------------------------
+
+#import <Foundation/Foundation.h>
+
+NS_ASSUME_NONNULL_BEGIN
+
+/*!
+ @protocol MSALTelemetryEventsObserving
```
Comment (oldalton) [inline MSAL/src/public/MSALTelemetryEventsObserving.h]: While I agree on the motivation of following better naming for the pattern, other SDKs all have dispatchers. Can you follow up with other SDKs and see what they think about renaming it?
If we cannot rename it in other SDKs, I'd prefer keeping dispatchers for consistency.

File: MSAL/src/public/MSALDefinitions.h [hunk 2]
```diff
@@ -117,5 +117,11 @@ typedef void (^MSALAccountsCompletionBlock)(NSArray<MSALAccount *> * _Nullable a
  */
 typedef void (^MSALLogCallback)(MSALLogLevel level, NSString * _Nullable message, BOOL containsPII);
 
+/*!
+ MSAL telemetry callback.
+ 
+ @param events Array of Dictionaries keyed on event name, valued on event data.
```
Comment (oldalton) [inline MSAL/src/public/MSALDefinitions.h]: If we only do aggregated data now, do we need to return an array of dictionaries, or a single dictionary should be sufficient?

File: MSAL/src/public/MSALTelemetry.h [hunk 3]
```diff
@@ -21,48 +21,50 @@
...
+@property (atomic) BOOL notifyOnFailureOnly DEPRECATED_MSG_ATTRIBUTE("use MSALGlobalConfig.telemetryConfig.notifyOnFailureOnly instead");
 
 /*!
- Remove a telemetry dispatcher added for receiving telemetry events.
- @param dispatcher An instance of MSALTelemetryDispatcher implementation added to the dispatches before.
+ If set NO, all telemetry events collected by MSAL will be dispatched;
+ If set YES, MSAL will dispatch only one event for each acquire token call.
  */
-- (void)removeDispatcher:(nonnull id<MSALTelemetryDispatcher>)dispatcher DEPRECATED_MSG_ATTRIBUTE("use MSALGlobalConfig.telemetryConfig removeDispatcher: instead");
+@property (atomic) BOOL aggregationRequired DEPRECATED_MSG_ATTRIBUTE("use MSALGlobalConfig.telemetryConfig.aggregationRequired instead");
```
Comment (oldalton) [inline MSAL/src/public/MSALTelemetry.h]: if we always do aggregation, should we get rid of this property altogether?

File: MSAL/src/public/MSALDefinitions.h (hunk 2 above)
Comment (antrix1989) [reply MSAL/src/public/MSALDefinitions.h]: Changed to singe dictionary.

File: MSAL/src/public/MSALTelemetry.h (hunk 3 above)
Comment (antrix1989) [reply MSAL/src/public/MSALTelemetry.h]: done

File: MSAL/src/public/configuration/global/MSALLoggerConfig.h [hunk 4]
```diff
@@ -35,11 +35,11 @@ NS_ASSUME_NONNULL_BEGIN
 /*!
  The minimum log level for messages to be passed onto the log callback.
  */
-@property MSALLogLevel logLevel;
+@property (atomic) MSALLogLevel logLevel;
 /*!
  MSAL provides logging callbacks that assist in diagnostics. There is a boolean value in the logging callback that indicates whether the message contains user information. If PiiLoggingEnabled is set to NO, the callback will not be triggered for log messages that contain any user information. By default the library will not return any messages with user information in them.
  */
-@property BOOL piiEnabled;
+@property (atomic) BOOL piiEnabled;
```
Comment (oldalton) [inline MSAL/src/public/configuration/global/MSALLoggerConfig.h]: do we actually ensure atomicity of those properties? Didn't we internally call MSIDTelemetry?

File: MSAL/src/telemetry/MSALAggregatedDispatcher.m [hunk 5]
```diff
@@ -17,24 +15,43 @@
...
-//------------------------------------------------------------------------------
 
-#import "MSALTestAppTelemetryDispatcher.h"
+#import "MSALAggregatedDispatcher.h"
+#import "MSIDDefaultDispatcher+Internal.h"
+#import "MSIDTelemetryBaseEvent.h"
+#import "MSIDTelemetryEventStrings.h"
+#import "MSALTelemetryEventsObservingProxy.h"
+
+@interface MSALAggregatedDispatcher ()
```
Comment (oldalton) [inline MSAL/src/telemetry/MSALAggregatedDispatcher.m]: Do we even need this class? Doesn't seem to be doing much. Can we instead call MSALTelemetryEventsObservingProxy's functionality directly?

File: MSAL/src/public/configuration/global/MSALLoggerConfig.h (hunk 4 above)
Comment (antrix1989) [reply MSAL/src/public/configuration/global/MSALLoggerConfig.h]: yes, we are calling MSIDLogger, but its property is atomic. Same case with MSALTelemetryConfig,

File: MSAL/src/telemetry/MSALAggregatedDispatcher.m (hunk 5 above)
Comment (antrix1989) [reply MSAL/src/telemetry/MSALAggregatedDispatcher.m]: Good point. Removed MSAL dispatchers.

**Pull request merged**